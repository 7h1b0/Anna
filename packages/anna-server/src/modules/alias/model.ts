import Ajv from 'ajv';
import { v4 as uuidv4 } from 'uuid';
import knex from '../../knexClient';
import aliasSchema from './schema';
import { omit } from '../../utils/utils';

export const TABLE = 'alias';
export const COLUMNS = [
  'aliasId',
  'sceneId',
  'name',
  'description',
  'enabled',
  'startTime',
  'endTime',
  'createdAt',
  'updatedAt',
  'createdBy',
];

type Alias = {
  aliasId: string;
  sceneId: string;
  name: string;
  description?: string;
  startTime?: number;
  endTime?: number;
  enabled: boolean;
  createdAt: number;
  updatedAt: number;
  createdBy: string;
};

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(aliasSchema, data);
}

export async function findAll(): Promise<Alias[]> {
  return knex(TABLE).select(COLUMNS);
}

export async function findById(aliasId: string): Promise<Alias> {
  return knex(TABLE).first(COLUMNS).where('aliasId', aliasId);
}

export async function findByName(name: string): Promise<Alias[]> {
  return knex(TABLE).select(COLUMNS).where('name', name);
}

export async function save({
  sceneId,
  name,
  description,
  createdBy,
  enabled = true,
  startTime,
  endTime,
}: {
  sceneId: string;
  name: string;
  description: string;
  createdBy: string;
  enabled?: boolean;
  startTime?: number;
  endTime?: number;
}): Promise<string> {
  const aliasId = uuidv4();
  await knex(TABLE).insert({
    aliasId,
    sceneId,
    description,
    name,
    enabled,
    createdBy,
    startTime,
    endTime,
  });
  return aliasId;
}

export function remove(aliasId: string) {
  return knex(TABLE).where('aliasId', aliasId).del();
}

export function findByIdAndUpdate(aliasId: string, payload: Partial<Alias>) {
  const safePayload = omit(
    payload,
    'aliasId',
    'createdAt',
    'updatedAt',
    'createdBy',
  );
  return knex(TABLE).update(safePayload).where('aliasId', aliasId);
}

export function isInteger(value: unknown | number): value is number {
  return Number.isInteger(value);
}

/**
 * Given aliases, returns one alias respecting the constraints
 */
export function resolveActiveAlias(aliases: Alias[]): Alias | undefined {
  const time = new Date().getHours();
  const activeAlias = aliases
    .filter((alias) => alias.enabled)
    .filter((alias) => {
      if (isInteger(alias.startTime) && isInteger(alias.endTime)) {
        if (alias.startTime < alias.endTime) {
          return time >= alias.startTime && time < alias.endTime;
        }
        return time >= alias.startTime || time < alias.endTime;
      }
      return true;
    });

  return activeAlias.pop();
}
