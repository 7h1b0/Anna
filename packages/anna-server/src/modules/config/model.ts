import knex from '../../knexClient';

export const TABLE = 'config';

export type Config = {
  isAway: boolean;
};

export async function fetchIsAway(): Promise<{ isAway: true }> {
  return knex(TABLE).first('isAway');
}

export function updateIsAway(isAway: boolean) {
  return knex(TABLE).update({ isAway });
}
