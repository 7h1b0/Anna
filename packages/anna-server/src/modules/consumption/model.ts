import knex from '../../knexClient';

export const TABLE = 'consumption';
export const COLUMNS = ['date', 'value'];

export type Consumption = {
  date: number;
  value: number;
};

export async function findLastEntries(limit: number): Promise<Consumption[]> {
  return knex(TABLE)
    .select(COLUMNS)
    .whereNotNull('value')
    .orderBy('date', 'desc')
    .limit(limit);
}

export async function findLastEntry(): Promise<Consumption> {
  return knex(TABLE)
    .first(COLUMNS)
    .whereNotNull('value')
    .orderBy('date', 'desc')
    .limit(1);
}

export async function save(consumptions: { date: Date; value: number }[]) {
  return knex.transaction((trx) => {
    return Promise.all(
      consumptions.map((consumption) => trx.insert(consumption).into(TABLE)),
    );
  });
}
