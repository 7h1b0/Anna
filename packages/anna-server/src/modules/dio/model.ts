import Ajv from 'ajv';
import knex from '../../knexClient';
import dioSchema from './schema';

export const TABLE = 'dios';

export type Dio = {
  dioId: number;
  roomId: string;
  name: string;
};

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(dioSchema, data);
}

export async function findAll(): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name');
}

export async function findById(dioId: number): Promise<Dio> {
  return knex(TABLE).first('dioId', 'roomId', 'name').where('dioId', dioId);
}

export async function findByRoomId(roomId: string): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name').where('roomId', roomId);
}

export async function save(newDio: Dio): Promise<Dio> {
  return knex(TABLE)
    .insert(newDio)
    .then(() => newDio);
}

export function remove(dioId: number) {
  return knex(TABLE).where('dioId', dioId).del();
}

export function findByIdAndUpdate(
  dioId: number,
  payload: Partial<Omit<Dio, 'dioId'>>,
) {
  return knex(TABLE).update(payload).where('dioId', dioId);
}
