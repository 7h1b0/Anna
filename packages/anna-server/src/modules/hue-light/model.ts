import knex from '../../knexClient';
export const TABLE = 'lights';
export const COLUMNS = ['roomId', 'lightId'];

export function findRoomId(lightId: number): Promise<string[]> {
  return knex(TABLE).first('roomId').where('lightId', lightId);
}

export function findAll() {
  return knex(TABLE).select(...COLUMNS);
}

export function findByRoomId(roomId: string): Promise<number[]> {
  return knex(TABLE)
    .select('lightId')
    .where('roomId', roomId)
    .then((res) => res.map(({ lightId }) => lightId));
}

export function remove(lightId: number): Promise<number> {
  return knex(TABLE).where('lightId', lightId).del();
}

export async function insertOrUpdate(lightId: number, roomId: string) {
  await knex.raw(
    'INSERT INTO lights (lightId, roomId) values (?, ?) ON DUPLICATE KEY UPDATE roomId=VALUES(roomId)',
    [lightId, roomId],
  );
}
