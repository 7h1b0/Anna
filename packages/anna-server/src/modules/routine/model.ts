import Ajv from 'ajv';
import { v4 as uuidv4 } from 'uuid';
import knex from '../../knexClient';
import routineSchema from './schema';
import {
  computeNextRunAt,
  isValidCron,
  isTimestamp,
} from '../../services/scheduleService';
import * as ScheduleService from '../../services/scheduleService';
import dispatch from '../../utils/dispatch';
import { omit } from '../../utils/utils';
import { fetchIsAway } from '../config/model';
import * as logger from '../../utils/logger';
import { callScene } from '../../utils/actions';

export const TABLE = 'routines';
export const COLUMNS = [
  'routineId',
  'name',
  'interval',
  'sceneId',
  'runAtBankHoliday',
  'runWhenAway',
  'enabled',
  'createdAt',
  'updatedAt',
  'lastRunAt',
  'nextRunAt',
  'createdBy',
  'lastStatus',
];

export class Routine {
  constructor(
    public routineId: string,
    public sceneId: string,
    public name: string,
    public interval: string,
    public nextRunAt: number | Date,
    public createdBy: string,
    public enabled = true,
    public runAtBankHoliday = true,
    public runWhenAway = false,
    public createdAt?: number | Date,
    public updatedAt?: number | Date,
    public lastRunAt?: number | Date,
    public lastStatus?: string,
  ) {}
}

export function updateAllNextRunAt(routines: Routine[]) {
  return knex.transaction(function (trx) {
    const reqs = routines.map((routine) => {
      const nextRunAt = computeNextRunAt(
        routine.interval,
        routine.runAtBankHoliday,
      );

      return knex(TABLE)
        .transacting(trx)
        .update({ nextRunAt })
        .where('routineId', routine.routineId);
    });

    Promise.all(reqs).then(trx.commit).catch(trx.rollback);
  });
}

export function validate(data: Partial<Routine>) {
  const ajv = new Ajv();
  const isValid = ajv.validate(routineSchema, data);
  if (isValidCron(data.interval)) {
    return isValid;
  } else if (isTimestamp(data.interval)) {
    return isValid && parseInt(data.interval ?? '', 10) > Date.now();
  }
  return false;
}

export async function findAll(): Promise<Routine[]> {
  return knex(TABLE).select(COLUMNS);
}

export async function findById(routineId: string): Promise<Routine> {
  return knex(TABLE).first(COLUMNS).where('routineId', routineId);
}

export async function save(
  createdBy: string,
  name: string,
  sceneId: string,
  interval: string,
  enabled?: boolean,
  runAtBankHoliday?: boolean,
  ddrunWhenAway?: boolean,
) {
  const routineId = uuidv4();
  const nextRunAt = computeNextRunAt(interval, runAtBankHoliday);

  const routine = new Routine(
    routineId,
    sceneId,
    name,
    interval,
    nextRunAt,
    createdBy,
    enabled,
    runAtBankHoliday,
    ddrunWhenAway,
  );

  await knex(TABLE).insert(routine);
  return routine;
}

export async function remove(routineId: string): Promise<number> {
  return knex(TABLE).where('routineId', routineId).del();
}

export async function findByIdAndUpdate(
  routineId: string,
  payload: Routine,
): Promise<number> {
  const nextRunAt = computeNextRunAt(
    payload.interval,
    payload.runAtBankHoliday,
  );

  const updatedPayload = Object.assign({}, payload, { nextRunAt });
  return knex(TABLE).update(updatedPayload).where('routineId', routineId);
}

export async function run(routine: Routine) {
  const done = (status: 'success' | 'failed' | 'skipped') => {
    const lastRunAt = new Date();

    const updatedRoutine = {
      ...omit(routine, 'createdAt', 'updatedAt'),
      lastRunAt,
      lastStatus: status,
    };

    return findByIdAndUpdate(routine.routineId, updatedRoutine);
  };

  try {
    const { runWhenAway } = routine;
    const isAway = await fetchIsAway();

    if (isAway && !runWhenAway) {
      return done('skipped');
    }

    await dispatch(callScene(routine.sceneId));
    return done('success');
  } catch (error) {
    logger.error(JSON.stringify(error));
    return done('failed');
  }
}

export function schedule(routine: Routine): void {
  ScheduleService.stop(routine.routineId); // Insure to stop process if already exists

  if (routine.enabled !== false) {
    ScheduleService.schedule(
      routine.routineId,
      routine.interval,
      () => run(routine),
      { runAtBankHoliday: routine.runAtBankHoliday },
    );
  }
}

export async function load() {
  const routines = await findAll();
  routines.map((routine) => schedule(routine));
  await updateAllNextRunAt(routines);
}
