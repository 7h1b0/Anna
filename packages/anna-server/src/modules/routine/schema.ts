export default {
  $id: 'routine',
  type: 'object',
  additionalProperties: false,
  properties: {
    name: {
      type: 'string',
      minLength: 3,
    },
    interval: { type: 'string' },
    sceneId: {
      type: 'string',
      pattern: '[a-z0-9-]{36}',
    },
    enabled: { type: 'boolean' },
    runAtBankHoliday: { type: 'boolean' },
  },
  required: ['name', 'sceneId', 'interval'],
};
