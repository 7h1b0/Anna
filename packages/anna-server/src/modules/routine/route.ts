import { Router } from 'express';
import * as Routine from '../../modules/routine/model';
import * as ScheduleService from '../../services/scheduleService';
import * as logger from '../../utils/logger';

const routes = Router();
routes
  .route('/api/routines')
  .get((req, res) => {
    Routine.findAll()
      .then((routines) => res.json(routines))
      .catch((err) => res.status(500).send({ err }));
  })
  .post((req, res) => {
    const isValid = Routine.validate(req.body);
    if (!isValid) {
      res.sendStatus(400);
    } else {
      const userId = 'undefined';
      const {
        name,
        sceneId,
        interval,
        enabled,
        runAtBankHoliday,
        runWhenAway,
      } = req.body;

      Routine.save(
        userId,
        name,
        sceneId,
        interval,
        enabled,
        runAtBankHoliday,
        runWhenAway,
      )
        .then((routine) => {
          Routine.schedule(routine);
          res.status(201).json({ routineId: routine.routineId });
        })
        .catch((err) => {
          res.status(500).send({ err });
        });
    }
  });

routes
  .route('/api/routines/:routineId([a-fA-F0-9-]{36})')
  .get((req, res) => {
    Routine.findById(req.params.routineId)
      .then((routine) => {
        if (!routine) {
          res.sendStatus(404);
        } else {
          res.json(routine);
        }
      })
      .catch((err) => {
        res.status(500).send({ err });
      });
  })
  .patch((req, res) => {
    const isValid = Routine.validate(req.body);
    if (!isValid) {
      res.sendStatus(400);
    } else {
      const routineId = req.params.routineId;
      Routine.findByIdAndUpdate(routineId, req.body)
        .then(async (rowsAffected) => {
          if (rowsAffected) {
            const routine = await Routine.findById(routineId);
            Routine.schedule(routine);
            res.sendStatus(204);
          } else {
            res.sendStatus(404);
          }
        })
        .catch((err) => res.status(500).send({ err }));
    }
  })
  .delete((req, res) => {
    Routine.remove(req.params.routineId)
      .then((removedScene) => {
        if (removedScene < 1) {
          res.sendStatus(404);
        } else {
          ScheduleService.stop(req.params.routineId);
          res.sendStatus(204);
        }
      })
      .catch((err) => {
        res.status(500).send({ err });
      });
  });

routes.get('/api/routines/:routineId([a-fA-F0-9-]{36})/action', (req, res) => {
  Routine.findById(req.params.routineId)
    .then((routine) => {
      if (routine) {
        return Routine.run(routine);
      }
      throw new Error('Routine not found');
    })
    .then(() => res.end())
    .catch((err) => {
      if (err.message === 'Routine not found') {
        res.sendStatus(404);
      } else {
        logger.error(err);
        res.status(500).send({ err });
      }
    });
});

export default routes;
