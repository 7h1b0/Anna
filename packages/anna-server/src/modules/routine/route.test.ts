import request from 'supertest';
import * as lolex from '@sinonjs/fake-timers';
import { setTimeout } from 'timers/promises';

import { createUser } from 'factories';
import knex from '../../knexClient';
import * as Routine from '../../modules/routine/model';
import * as User from '../../modules/user/model';
import app from '../../index';
import dispatch from '../../utils/dispatch';

jest.mock('../../utils/dispatch');

const user = createUser({ userId: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f' });
const useraway = createUser({
  userId: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879g',
  token: 'user_away_token',
  isAway: true,
});
const initRoutines = [
  {
    routineId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    sceneId: '00c1d78e-fd1c-4717-b610-65d2fa3d01b2',
    name: 'test at 5am',
    interval: '0 5 * * *',
    enabled: true,
    runAtBankHoliday: true,
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
  {
    routineId: '1fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    sceneId: '10c1d78e-fd1c-4717-b610-65d2fa3d01b2',
    name: 'test at 9am',
    interval: '0 9 * * *',
    enabled: true,
    runAtBankHoliday: false,
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
  {
    routineId: '2fc1d78e-fd1c-4717-b610-65d2fa3d01b3',
    sceneId: '10c1d78e-fd1c-4717-b610-65d2fa3d01b2',
    name: 'test at 12am',
    interval: '0 0 * * *',
    enabled: false,
    runAtBankHoliday: false,
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
];

describe('Routine API', () => {
  beforeAll(async () => {
    await knex(User.TABLE).truncate();
    await knex(User.TABLE).insert([user, useraway]);
  });

  beforeEach(async () => {
    // @ts-expect-error dispatch is a mock
    dispatch.mockClear();
    await knex(Routine.TABLE).truncate();
    await knex(Routine.TABLE).insert(initRoutines);
  });

  describe('/api/routines', () => {
    describe('GET', () => {
      it('should retun all routines', async () => {
        const response = await request(app)
          .get('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(200);
        expect(response.body).toMatchSnapshot();
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });
    });

    describe('POST', () => {
      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });

      it('should retun 400 when routine is invalid', async () => {
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'room_updated',
          });

        expect(response.status).toBeBadRequest();
      });

      it('should retun 400 when interval is invalid', async () => {
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'room_updated',
            interval: '30* * * * *',
          });

        expect(response.status).toBeBadRequest();
      });

      it('should save and start a new routine', async () => {
        const clock = lolex.install({ now: new Date('2017-08-12T16:00:15') });
        const payload = {
          sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
          name: 'new_routine',
          interval: '30 * * * * *',
        };
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send(payload);

        clock.next();
        clock.uninstall();

        // FIXME: Dispatch needs to fetch user status before calling dispatch
        await setTimeout(1000);
        expect(dispatch).toHaveBeenCalledTimes(1);
        expect(dispatch).toHaveBeenCalledWith({
          type: 'SCENE',
          targetId: payload.sceneId,
        });
        expect(response.status).toHaveStatusOk();

        const routine = await knex(Routine.TABLE)
          .first()
          .where('routineId', response.body.routineId);

        expect(routine).toMatchSnapshot({
          routineId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
          createdAt: expect.any(Number),
          updatedAt: expect.any(Number),
          nextRunAt: expect.any(Number),
          lastRunAt: expect.any(Number),
        });
      });

      it('should skip a routine when user is away', async () => {
        const clock = lolex.install({ now: new Date('2017-08-12T16:00:15') });
        const payload = {
          sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b3',
          name: 'new_routine',
          interval: '30 * * * * *',
        };
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', useraway.token)
          .send(payload);

        clock.next();
        clock.uninstall();

        expect(response.status).toHaveStatusOk();
        expect(dispatch).not.toHaveBeenCalled();

        // FIXME: wait for routine run to update the object
        await setTimeout(1000);
        const routine = await knex(Routine.TABLE)
          .first()
          .where('routineId', response.body.routineId);

        expect(routine.lastStatus).toBe('skipped');
      });

      it('should allow to run a routine even is user is away', async () => {
        const clock = lolex.install({ now: new Date('2017-08-12T16:00:15') });
        const payload = {
          sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b3',
          name: 'new_routine',
          interval: '30 * * * * *',
          runWhenUserIsAway: true,
        };
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', useraway.token)
          .send(payload);

        clock.next();
        clock.uninstall();

        expect(response.status).toHaveStatusOk();

        // FIXME: Dispatch needs to fetch user status before calling dispatch
        await setTimeout(1000);
        expect(dispatch).toHaveBeenCalledTimes(1);
        expect(dispatch).toHaveBeenCalledWith({
          type: 'SCENE',
          targetId: payload.sceneId,
        });
      });

      it('should create a temporary routine', async () => {
        const clock = lolex.install({ now: new Date('2017-08-12T16:00:00') });
        const payload = {
          sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01dd',
          name: 'new_routine',
          interval: `${Date.now() + 2000}`,
        };
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send(payload);

        expect(response.status).toHaveStatusOk();

        await clock.nextAsync();

        // FIXME: Dispatch needs to fetch user status before calling dispatch
        await setTimeout(1000);
        expect(dispatch).toHaveBeenCalledTimes(1);
        expect(dispatch).toHaveBeenCalledWith({
          type: 'SCENE',
          targetId: payload.sceneId,
        });
        expect(clock.countTimers()).toBeFalsy();

        clock.uninstall();
      });

      it('should reject temporary routine in the past', async () => {
        const clock = lolex.install({ now: new Date('2017-08-12T16:00:00') });
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'new_routine',
            interval: `${Date.now() - 10000}`,
          });

        expect(response.status).toBeBadRequest();
        expect(clock.countTimers()).toBeFalsy();

        clock.uninstall();
      });
    });
  });

  describe('/api/routines/:routineId', () => {
    describe('GET', () => {
      it('should return one routine', async () => {
        const response = await request(app)
          .get(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.body).toMatchSnapshot();
      });

      it('should return 404 if routineId is unknow', async () => {
        const response = await request(app)
          .get('/api/routines/faaed78e-fd1c-4717-b610-65d2fa3d01b2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
        expect(response.body).toEqual({});
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
        expect(response.body).toEqual({});
      });
    });

    describe('PATCH', () => {
      it('should return 404 if routineId is unknow', async () => {
        const response = await request(app)
          .patch('/api/routines/faaed78e-fd1c-4717-b610-65d2fa3d01b2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'room_updated',
            interval: '* * * * * *',
          });

        expect(response.status).toBe(404);
        expect(response.body).toEqual({});
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .patch(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake')
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'room_updated',
            interval: '* * * * * *',
          });

        expect(response.status).toBeUnauthorized();
        expect(response.body).toEqual({});
      });

      it('should return 400 when routine is invalid', async () => {
        const response = await request(app)
          .patch(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'room_updated',
          });

        expect(response.status).toBeBadRequest();
      });

      it('should return 400 when interval is invalid', async () => {
        const response = await request(app)
          .patch(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'room_updated',
            interval: '05* * *',
          });

        expect(response.status).toBeBadRequest();
      });

      it('should disable routine', async () => {
        const clock = lolex.install({ now: new Date('2017-08-12T16:00:00') });

        await Routine.load();
        expect(clock.countTimers()).toBe(2); // routines 0 and 1 should run

        const response = await request(app)
          .patch(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'routine_updated',
            interval: '0 5 * * *',
            enabled: false,
          });

        expect(response.status).toHaveStatusOk();
        expect(clock.countTimers()).toBe(1); // routine 0 should be disabled

        clock.next();
        expect(Date.now()).toBe(new Date('2017-08-13T09:00:00').getTime());

        // FIXME: Dispatch needs to fetch user status before calling dispatch
        await setTimeout(1000);
        expect(dispatch).toHaveBeenCalledTimes(1);
        clock.uninstall();

        const routine = await knex(Routine.TABLE)
          .first()
          .where('routineId', initRoutines[0].routineId);

        expect(routine.enabled).toBeFalsy();
      });

      it('should enable routine and compute nextRunAt', async () => {
        const clock = lolex.install({ now: new Date('2017-08-12T16:00:00') });

        const response = await request(app)
          .patch(`/api/routines/${initRoutines[2].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'routine_updated',
            interval: '0 12 * * *',
            enabled: true,
          });

        expect(response.status).toHaveStatusOk();
        expect(clock.countTimers()).toBe(1);

        const routine = await knex(Routine.TABLE)
          .first()
          .where('routineId', initRoutines[2].routineId);

        expect(routine.nextRunAt).toEqual(
          new Date('2017-08-13T12:00:00').getTime(),
        );
        expect(routine).toMatchSnapshot({
          createdAt: expect.any(Number),
          updatedAt: expect.any(Number),
          nextRunAt: expect.any(Number),
        });

        clock.uninstall();
      });
    });

    describe('DELETE', () => {
      it('should return 404 if routineId is unknow', async () => {
        const response = await request(app)
          .delete('/api/routines/faaed78e-fd1c-4717-b610-65d2fa3d01b2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
        expect(response.body).toEqual({});
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .delete(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
        expect(response.body).toEqual({});
      });

      it('should delete and stop a routine', async () => {
        const clock = lolex.install();

        const response = await request(app)
          .delete(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toHaveStatusOk();
        expect(clock.countTimers()).toBe(0);
        clock.uninstall();

        const routines = await knex(Routine.TABLE).select('routineId');
        expect(routines).toHaveLength(2);
      });
    });
  });

  describe('/api/routines/:routineId/action', () => {
    it('should run the routine', async () => {
      const runSpy = jest.spyOn(Routine, 'run');
      const response = await request(app)
        .get(`/api/routines/${initRoutines[0].routineId}/action`)
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toHaveStatusOk();
      expect(response.body).toEqual({});
      expect(Routine.run).toHaveBeenCalled();

      expect(runSpy.mock.calls).toMatchSnapshot();
      runSpy.mockRestore();
    });

    it('should return 404 if routineId is unknow', async () => {
      const response = await request(app)
        .get('/api/routines/faaed78e-fd1c-4717-b610-65d2fa3d01b2/action')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(404);
      expect(response.body).toEqual({});
    });

    it('should return 401 when user is not authenticated', async () => {
      const response = await request(app)
        .get(`/api/routines/${initRoutines[0].routineId}/action`)
        .set('Accept', 'application/json')
        .set('x-access-token', 'fake');

      expect(response.status).toBeUnauthorized();
      expect(response.body).toEqual({});
    });
  });
});
