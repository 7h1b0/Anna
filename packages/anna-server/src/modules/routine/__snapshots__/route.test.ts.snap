// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routine API /api/routines GET should retun all routines 1`] = `
Array [
  Object {
    "createdAt": 1514764800000,
    "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "enabled": true,
    "interval": "0 5 * * *",
    "lastRunAt": 0,
    "lastStatus": null,
    "name": "test at 5am",
    "nextRunAt": 0,
    "routineId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "runAtBankHoliday": true,
    "runWhenAway": false,
    "sceneId": "00c1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "updatedAt": 1514851200000,
  },
  Object {
    "createdAt": 1514764800000,
    "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "enabled": true,
    "interval": "0 9 * * *",
    "lastRunAt": 0,
    "lastStatus": null,
    "name": "test at 9am",
    "nextRunAt": 0,
    "routineId": "1fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "runAtBankHoliday": false,
    "runWhenAway": false,
    "sceneId": "10c1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "updatedAt": 1514851200000,
  },
  Object {
    "createdAt": 1514764800000,
    "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "enabled": false,
    "interval": "0 0 * * *",
    "lastRunAt": 0,
    "lastStatus": null,
    "name": "test at 12am",
    "nextRunAt": 0,
    "routineId": "2fc1d78e-fd1c-4717-b610-65d2fa3d01b3",
    "runAtBankHoliday": false,
    "runWhenAway": false,
    "sceneId": "10c1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "updatedAt": 1514851200000,
  },
]
`;

exports[`Routine API /api/routines POST should save and start a new routine 1`] = `
Object {
  "createdAt": Any<Number>,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "enabled": true,
  "interval": "30 * * * * *",
  "lastRunAt": Any<Number>,
  "lastStatus": "success",
  "name": "new_routine",
  "nextRunAt": Any<Number>,
  "routineId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "runAtBankHoliday": true,
  "runWhenAway": false,
  "sceneId": "faaed78e-fd1c-4717-b610-65d2fa3d01b2",
  "updatedAt": Any<Number>,
}
`;

exports[`Routine API /api/routines/:routineId GET should return one routine 1`] = `
Object {
  "createdAt": 1514764800000,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "enabled": true,
  "interval": "0 5 * * *",
  "lastRunAt": 0,
  "lastStatus": null,
  "name": "test at 5am",
  "nextRunAt": 0,
  "routineId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  "runAtBankHoliday": true,
  "runWhenAway": false,
  "sceneId": "00c1d78e-fd1c-4717-b610-65d2fa3d01b2",
  "updatedAt": 1514851200000,
}
`;

exports[`Routine API /api/routines/:routineId PATCH should enable routine and compute nextRunAt 1`] = `
Object {
  "createdAt": Any<Number>,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "enabled": true,
  "interval": "0 12 * * *",
  "lastRunAt": 0,
  "lastStatus": null,
  "name": "routine_updated",
  "nextRunAt": Any<Number>,
  "routineId": "2fc1d78e-fd1c-4717-b610-65d2fa3d01b3",
  "runAtBankHoliday": false,
  "runWhenAway": false,
  "sceneId": "faaed78e-fd1c-4717-b610-65d2fa3d01b2",
  "updatedAt": Any<Number>,
}
`;

exports[`Routine API /api/routines/:routineId/action should run the routine 1`] = `
Array [
  Array [
    RowDataPacket {
      "createdAt": 1514764800000,
      "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
      "enabled": true,
      "interval": "0 5 * * *",
      "lastRunAt": 0,
      "lastStatus": null,
      "name": "test at 5am",
      "nextRunAt": 0,
      "routineId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
      "runAtBankHoliday": true,
      "runWhenAway": false,
      "sceneId": "00c1d78e-fd1c-4717-b610-65d2fa3d01b2",
      "updatedAt": 1514851200000,
    },
  ],
]
`;
